aten/src/ATen/native/cuda/ForeachTensorAddScalar.cu

```diff
@@ -0,0 +1,96 @@
+// of a __device__ lambda not have internal linkage.
+
+namespace at { namespace native {
+
+namespace {
+
+template<typename x_t, typename out_t>
+struct AddScalarFunctor {
+    __device__ void operator() (
+        int chunk_size,
+        TensorListMetadata<2>& tl,
+        x_t scalar) {
+            int tensor_loc = tl.block_to_tensor[blockIdx.x];
+            int chunk_idx = tl.block_to_chunk[blockIdx.x];
+            int n = tl.sizes[tensor_loc];
+
+            x_t* x = (x_t*)tl.addresses[0][tensor_loc];
+            x += chunk_idx * chunk_size;
+
+            out_t* out = (out_t*)tl.addresses[1][tensor_loc];
+            out += chunk_idx * chunk_size;
+
+            n -= chunk_idx * chunk_size;
+
+            x_t r_x[kILP];
+            out_t r_out[kILP];
+
+            // to make things simple, we put aligned case in a different code path
+            if(n % kILP == 0 && chunk_size % kILP == 0 && is_aligned(x) && is_aligned(out)) {
+                for(int i_start = threadIdx.x; i_start * kILP < n && i_start * kILP < chunk_size; i_start += blockDim.x) {
+                    // load
+                    load_store(r_x, x, 0 , i_start);
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<x_t>(r_x[ii]) + scalar;
+                    }
+                    // store
+                    load_store(out, r_out, i_start, 0);
+                }
+            }
+            else {
+                // Non-divergent exit condition for __syncthreads, not necessary here
+                for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = 0;
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size) {
+                            r_x[ii] = x[i];
+                        }
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<x_t>(r_x[ii]) + scalar;
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size)
+                            out[i] = r_out[ii];
+                    }
+                }
+            }
+        }
+};
+
+} // namespace
+
+std::vector<Tensor> foreach_tensor_add_scalar_kernel_cuda(TensorList tensors, Scalar scalar) {
+    TORCH_CHECK(tensors.size() > 0, "Tensor list must have at least one tensor.");
+
+    if (!check_fast_route(tensors, scalar)) {
+        return at::native::foreach_add_scalar_kernel_cpu(tensors, scalar);
+    }
+
+    std::vector<std::vector<at::Tensor>> tensor_lists; 
+    std::vector<at::Tensor> vec_res;
+    for (int i = 0; i < tensors.size(); i++) {
```

Comment:
nit: `for (const auto& t: tensors)` - what if you list is so large integer overflows?